### PR TITLE
feat(CustomFields): Add custom fields feature

### PR DESCRIPTION
**Pull request by gabriele-v**

### CODE REVIEW

File: src/customfielddialog.cpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+void mmCustomFieldDialog::OnMove(const wxPoint & RefPos, const wxSize & RefSize)
+{
+    this->Move(wxPoint(RefPos.x + RefSize.GetWidth(), RefPos.y));
+}
+
+void mmCustomFieldDialog::OnSave(const bool OpenStatus)
+{
+    if (m_RefreshRequested)
+    {
+        Destroy();
```
Comment (vomikan) [inline src/customfielddialog.cpp]: Detected memory leaks!
How to reproduce:
Create new transaction by press the '+' ico in tool bar. Then immediatly press the "Close" button. Close the app.

File: src/model/Model_CustomField.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    for (const auto& item : FIELDTYPE_CHOICES) types.Add(item.second);
+    return types;
+}
+
+wxString Model_CustomField::getTooltip(const wxString& Properties)
+{
+    json::Object jsonProperties;
+    std::wstringstream jsonPropertiesStream;
+
+    jsonPropertiesStream << Properties.ToStdWstring();
```
Comment (vomikan) [inline src/model/Model_CustomField.cpp]: To start working with this branch I've thanged it to :
~~~~~
jsonPropertiesStream << (Properties.empty() ? L"{}" : Properties.ToStdWstring());
~~~~~
for any jsonPropertiesStream in this file.
In other case the Error message appears: Unexpected end o ftocken string.

Comment (gabriele-v) [inline src/model/Model_CustomField.cpp]: As already said this isn't needed in production. You have used a very alpha implementation that has a different DB structure. If you want to fix excute this query on your DB:
UPDATE CUSTOMFIELD_V1 SET PROPERTIES = '{}' WHERE PROPERTIES = ''

File: src/customfielddialog.cpp (hunk 1 above)
Comment (gabriele-v) [inline src/customfielddialog.cpp]: Cannot reproduce with latest Stefan fixes: https://github.com/gabriele-v/moneymanagerex

Comment (gabriele-v) [inline src/customfielddialog.cpp]: PS: How to update pull request?

File: src/model/Model_CustomField.cpp (hunk 2 above)
Comment (vomikan) [inline src/model/Model_CustomField.cpp]: Stefan's path does not help here for me.

**Comment by stef145g** [comment]
I have realigned my custom_fields branch to @gabriele-v master for testing:
now mmex will crash when I do the following:
1. edit a transaction (or New)
2. open the custom field dialog
3. close the custom dialog
4. Enter the transaction using OK...

**Comment by vomikan** [comment]
When I try to  open the custom field dialog:

 [image: Встроенное изображение 1]

2016-01-25 17:27 GMT+03:00 Stefano <notifications@github.com>:

**Comment by stef145g** [comment]
I cannot see the image as suggested above.

**Comment by vomikan** [comment]
I've attached the image.

2016-01-25 17:48 GMT+03:00 Stefano <notifications@github.com>:

File: src/model/Model_CustomField.cpp (hunk 2 above)
Comment (vomikan) [inline src/model/Model_CustomField.cpp]: Thank you. The update has fixed the issue.

2016-01-25 15:46 GMT+03:00 Gabriele-V <notifications@github.com>:

File: src/customfielddialog.cpp (hunk 1 above)
Comment (vomikan) [inline src/customfielddialog.cpp]: Close it first then commit again.

2016-01-25 16:10 GMT+03:00 Gabriele-V <notifications@github.com>:

**Comment by gabriele-v** [comment]
@stef145g Should be fixed now, it's a way that I don't like (hide frame instead destroy) but I haven't found another solution that won't add lot of redundant code to parent windows (transaction now, but other dialogs in future too)..

@vomikan I can't see image too, could you post directly to GitHub?

**Comment by stef145g** [comment]
@gabriele-v Your fix works.
Hiding frame is fine, as it will be destroyed when the transaction dialog is closed, regardless of the number of frames created.

I have discovered more places where it crashes and a fix provided in https://github.com/stef145g/moneymanagerex/tree/custom_fields branch for your evaluation.

**Comment by vomikan** [comment]
@stef145g From my point of view, better to clone  Gabriele's repo then sent PR to it.

**Comment by stef145g** [comment]
pull request created: https://github.com/gabriele-v/moneymanagerex/pull/1

**Comment by vomikan** [comment]
@stef145g & @gabriele-v guys if this PR stable enought - merge!

**Pull request merged**